go-controller/pkg/network-attach-def-controller/network_attach_def_controller_test.go

```diff
@@ -545,6 +551,7 @@ func TestSyncAll(t *testing.T) {
 					netconf: &network_A_Copy,
 				},
 			},
+			syncAllError: ErrNetworkControllerTopologyNotManaged,
```

Comment:
not sure if you got my question..
which bit of code is being tested here, the mocked up test code i.e "func (tncm *testNetworkControllerManager) NewNetworkController"
OR
the original code which is what I want testing around:
```
// NewNetworkController implements the networkAttachDefController.NetworkControllerManager
// interface function. This function is called by the net-attach-def controller when
// a secondary network is created.
func (sncm *secondaryNetworkClusterManager) NewNetworkController(nInfo util.NetInfo) (nad.NetworkController, error) {
	if !sncm.isTopologyManaged(nInfo) {
		return nil, nad.ErrNetworkControllerTopologyNotManaged
	}
```
? which newNetworkController manager are you testing here, cause you already added tests for the secondary bits above which are testing the real code right? which part is being tested here?

Reply:
I'm trying to explain to the best of my abilities:
- the error is thrown by the mocked `testNetworkControllerManager` . 
- the error is handled by the *real* `networkManagerImpl` , which is the component that ignores that error, which is what I'm trying to test - i.e. the code introduced in https://github.com/ovn-kubernetes/ovn-kubernetes/commit/a5ac6a33472df0aceda72eeecf737db18ef42e85

This is the flow the test goes through (checked with an attached debugger):
- nad controller start func (should succeed, but would fail without the code introduced in the aforementioned commit): https://github.com/ovn-kubernetes/ovn-kubernetes/blob/8afc23e6c84e255e625d2715dfc2dbbf7e56473e/go-controller/pkg/network-attach-def-controller/network_attach_def_controller_test.go#L613
- nad controller network manager is started: https://github.com/ovn-kubernetes/ovn-kubernetes/blob/8b8f353c5d3abd40902660287e32a9f4e5051977/go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go#L140
- the nad controller network manager implementation sync all function is invoked: https://github.com/ovn-kubernetes/ovn-kubernetes/blob/a5ac6a33472df0aceda72eeecf737db18ef42e85/go-controller/pkg/network-attach-def-controller/network_manager.go#L78
- the patched code (tolerate the error my mock thrown) is triggered: https://github.com/ovn-kubernetes/ovn-kubernetes/blob/a5ac6a33472df0aceda72eeecf737db18ef42e85/go-controller/pkg/network-attach-def-controller/network_manager.go#L218

Reply:
ack this makes total sense, thanks for the nice explanation, the calls are not intuitive at all - for some reason I kept thinking you are re-testing the "is the error triggered correctly or not" instead of which you are testing if its caught correctly or not here.

Reply:
I'd say the boundaries between a network manager and a network controller manager are a bit blurry for the casual reader.

I have to re-learn this every time I venture into this part of the code ...

Thanks for the review !